features/netsocket/nsapi_dns.h

```diff
@@ -77,6 +77,19 @@ nsapi_error_t nsapi_dns_query(NetworkStack *stack, const char *host,
         SocketAddress *addr, nsapi_version_t version = NSAPI_IPv4);
 
 /** Query a domain name server for an IP address of a given hostname
+ *
+ *  @param stack    Network stack as target for DNS query
+ *  @param host     Hostname to resolve
+ *  @param callback Callback that is called for result*
+ *  @param data     Caller defined data returned in callback
+ *  @param version  IP version to resolve (defaults to NSAPI_IPv4)
+ *  @return         0 on success, negative error code on failure
+ *                  NSAPI_ERROR_DNS_FAILURE indicates the host could not be found
+ */
+nsapi_error_t nsapi_dns_query_async(NetworkStack *stack, const char *host,
+        NetworkStack::hostbyname_cb_t callback, void *data, nsapi_version_t version = NSAPI_IPv4);
```

Comment:
This should be a `Callback` - that's slightly more flexible.

See https://os.mbed.com/docs/v5.7/reference/platform.html#callbacks

Reply:
Keep in mind this layer also supports C. But it would be useful to have a C++ overload. The callback class has a thunk member function that can be used when passing to C layer:

``` cpp
void hi(void (*cb)(void *data, int arg), void *data);

Callback<void(int)> cb = function_here;
hi(&Callback<void(int)>::thunk, &cb);
```